### ISSUE TITLE
[Bug]: expo-updates crashes with libsql

### ISSUE THREAD

**Issue by ospfranco** [OP-Engineering/op-sqlite#214]
### What happened?

Re-opening to avoid more tickets being created:

https://github.com/expo/expo/issues/33644

There seems to be some conflict on how libsql and expo-updates handle one the sqlite3 variables. This in turn crashes the app on start. The problem lies not in op-sqlite but rather in libsql itself. I've tried reaching out to Turso itself to see what can be done, but gotten no response.

### op-sqlite version

-

### React Native version

-

### Reproducible Example

-

**Comment by ospfranco** [comment]
I just carefully re-read everything and went through the expo-updates code. Expo-updates calls `sqlite3_open` (which internally calls `sqlite3_initialize`) and libsql doesn't like this. libsql contains this crashing assertion because it sets the configuration to SERIALIZED and also turns on the mutex for thread safety. There is nothing "wrong" expo-updates is doing, it's just libsql not being happy with sqlite3 being initialized before it can set a configuration.

One solution I would like to try is to open a connection before expo-updates does anything, but not sure how the expo tool chain works, if it loads it's own modules before the other modules. I guess one solution would be to delay the initialization of expo-updates but I don't see any API for it. Looking at the example by @jmatsushita I guess expo router (to I don't know how it works) is now the default so I don't see an index.js anymore where one could force the loading of op-sqlite ahead of time and create a dummy DB.

There might be an update on libsql that gets rid of the assertion, but libsql sets SERIALIZED mode and relies on the internal sqlite mutex to guarantee thread safety.

**Comment by ospfranco** [comment]
Yep, opening a database before expo gets to do anything works. It will require a plugin (which I will not write myself, whomever wants it should open a PR) or prebuilding the expo app.

![Image](https://github.com/user-attachments/assets/3581018b-39d9-405d-8e80-a96f75084477)

**Comment by ospfranco** [comment]
@jmatsushita I'm trying to use your example here

https://github.com/jmatsushita/expo-libsql-bug

I had to remove a bunch of dependencies which somehow didn't compile on Android but after getting the app to compile now I'm getting

![Image](https://github.com/user-attachments/assets/4ad63c7f-3a6a-4dd4-b9fe-9b6cb6ec8af9)

Could you give me a hand and fix the example so that it compiles and see if it crashes with the same conflict on Android.

**Comment by ospfranco** [comment]
@jmatsushita Does this even happen in Android? The example seems to compile and start just fine

**Comment by ospfranco** [comment]
Also looking at the js code, there is an error: `open` is not a promise, it's sync.

**Comment by jmatsushita** [comment]
Hi @ospfranco,

Thanks for looking into this. I have set aside the libsql backend of op-sqlite in my project for now (in part because of the lack of reactive queries [for the foreseeable future](https://github.com/OP-Engineering/op-sqlite/issues/205) and because of the closed source sync server for offline writes) so I'm afraid I don't really have time to help further. However FYI I only tested on iOS and not android. 

> I don't see an index.js anymore where one could force the loading of op-sqlite ahead of time and create a dummy DB.

AFAIK you can do this [in the root layout](https://docs.expo.dev/router/advanced/root-layout/) and it's also possible to use an `index.ts` like in the good old days with `import "expo-router/entry";`  if you DONT use the recommended `"main": "expo-router/entry"` and instead have `"main": "index"` in your `package.json` (I don't think that's really supported anymore but I'm not sure).

> Also looking at the js code, there is an error: open is not a promise, it's sync.
 
Ah yeah. Oops.

Great to see that there's now a documented workaround 🚀 

Cheers,

Jun
